Docs 5156 objectstores v4.3 cv (#1467)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/mule-object-stores.adoc

= Store Application Data Using Object Stores
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: mule, object, store

> An object store is a facility for storing objects in or across Mule applications. Mule runtime engine(Mule) uses object stores to persist data for eventual retrieval. Internally, Mule uses object stores in various filters, routers, and other message processors that need to store states between messages.

Reviewer: ```suggestion
An object store is a facility for storing objects in or across Mule applications. Mule runtime engine (Mule) uses object stores to persist data for eventual retrieval. Internally, Mule uses object stores in various filters, routers, and other message processors that need to store states between messages.
```

> If you deploy your application to CloudHub, you can also use Object Store V2.

Reviewer: "also" in addition to what?

== Use Cases

You can configure a custom object store to use in the following scenarios:

* When configuring an Idempotent Message Validator, to store the unique IDs in the specified object store.

> * When configuring a Redelivery policy in a listener(HTTP, File, JMS, VM, Socket), to store the redelivery count in the specified object store.

Reviewer: ```suggestion
* When configuring a Redelivery policy in a listener (HTTP, File, JMS, VM, Socket), to store the redelivery count in the specified object store.
```

> * When configuring a Cache Scope that uses a custom caching strategy, to store the cached data in the specified object store.

Reviewer: I tried to look up Cache Scope and got only "cache scope" and "Cache scope." 
Since you precede it with "a", I think we want "a" generic thing, not "the" named thing, and so I suggest lowercase.
Please let me know if this is wrong, for the Term Presentation List?
```suggestion
* When configuring a cache scope that uses a custom caching strategy, to store the cached data in the specified object store.
```

Reviewer 2: https://docs.mulesoft.com/mule-runtime/4.3/cache-scope "Cache scope"

* When a custom component uses an object store to persist information.

> * When storing or retrieving information from a Mule flow, by using the Object Store Connector.

Reviewer: ```suggestion
* When storing or retrieving information from a Mule flow, by using Anypoint Connector for Object Store (Object Store Connector).
```

> == Configuration

Reviewer: Object Store Configuration?
Object Store Configuration Choices?

By default, each Mule application has an object store that is persistent and is always available to the app. You can also create custom object stores, which can be persistent or in-memory.

> You can configure a custom object store for a specific component that supports its usage, or configure the object store as a global element, to be accessible for any component in your app.

Reviewer: The order here (specific component, global) does not match the order of the subsequent subtopics.
Either change one or the other so that the intro and topics list in the same order.

=== Configure a Global Object Store

To configure a global object store that can be referenced by other components in your application, follow these steps in Anypoint Studio:

. Open the *Global Elements* tab.
. Click *Create*.
. Select *Component configurations* > *Object Store*.
. Complete the required configuration attributes and click *OK*. +
image:mruntime-object-store-global.png[global object store]

Alternatively, you can create a global object store by including the following code as a top-level element in your application XML:

[source,xml,linenums]
----
<os:object-store name="myObjectStore"
   entryTtl="1"
   entryTtlUnit="HOURS"
   maxEntries="100"
   persistent="true"
   expirationInterval="30"
   expirationIntervalUnit="MINUTES" />
----

Configure each attribute with the desired value. +
See xref:object-store-connector::object-store-connector-reference.adoc#objectStore[Object Store Reference] for a definition of each XML attribute.

== Configure an Object Store for a Specific Component

> You can configure an object store that is available only for a specific component. For example, to configure an object store for an Idempotent Message Validator, follow these steps in Anypoint Studio:

Reviewer: Would you also confirm the capitalization of Idempotent Message Validator?
If not specific to MuleSoft, I suggest lowercase.

Reviewer 2: https://docs.mulesoft.com/mule-runtime/4.3/idempotent-message-validator "Idempotent Message Validator" component. 

All components: https://docs.mulesoft.com/mule-runtime/4.3/about-components

. Open the *Idempotent Message Validator* properties.
. In *Object Store*, select *Edit Inline*.
. Complete the required configuration attributes and click *OK*. +
image:mruntime-object-store-inline.png[Inline object store]

> Alternatively, you can create the object store by including the object store configuration inside the component element in the XML configuration file, for example:

Reviewer: ```suggestion
Alternatively, you can create the object store by including the object store configuration inside the component element in the XML configuration file:
```

[source,xml,linenums]
----
<idempotent-message-validator doc:name="Idempotent Message Validator" >
  <os:private-object-store
    maxEntries="20"
    entryTtl="20"
    entryTtlUnit="MILLISECONDS"
    expirationInterval="20"
    expirationIntervalUnit="MILLISECONDS"
    config-ref="ObjectStore_Config" />
</idempotent-message-validator>
----

== Manage Data Using the Object Store Connector

The Object Store connector enables you to perform different operations to interact with the object stores in your Mule application. Example operations include Store, Retrieve and Remove.

> See the xref:object-store-connector::index.adoc[Object Store connector] documentation for details about available operations and configuration examples.

Reviewer: ```suggestion
See the xref:object-store-connector::index.adoc[Object Store Connector] documentation for details about available operations and configuration examples.
```

== See Also

* xref:object-store-connector::object-store-to-define-a-new-os.adoc[Examples: Defining Object Stores]
* xref:object-store::index.adoc[Object Store V2]
* xref:idempotent-message-validator.adoc[Idempotent Message Validator]

> * xref:redelivery-policy.adoc[Redelivery policy]

Reviewer: ```suggestion
* xref:redelivery-policy.adoc[Redelivery Policy]
```
